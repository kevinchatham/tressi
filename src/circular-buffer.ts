
/**
 * A memory-efficient circular buffer implementation for storing a fixed number of items.
 * When the buffer is full, new items overwrite the oldest ones.
 */
export class CircularBuffer<T> {
  private buffer: T[];
  private capacity: number;
  private head = 0;
  private tail = 0;
  private isFull = false;

  /**
   * Creates a new CircularBuffer instance.
   * @param capacity The maximum number of items the buffer can hold.
   */
  constructor(capacity: number) {
    this.capacity = capacity;
    this.buffer = new Array<T>(capacity);
  }

  /**
   * Adds an item to the buffer. If the buffer is full, it overwrites the oldest item.
   * @param item The item to add.
   */
  add(item: T): void {
    this.buffer[this.tail] = item;
    this.tail = (this.tail + 1) % this.capacity;

    if (this.isFull) {
      this.head = (this.head + 1) % this.capacity;
    } else if (this.tail === this.head) {
      this.isFull = true;
    }
  }

  /**
   * Returns all the items in the buffer.
   * @returns An array containing the items.
   */
  getAll(): T[] {
    const result: T[] = [];
    let i = this.head;
    const end = this.tail;

    if (this.isFull) {
      // If full, loop from head to tail covering all elements
      do {
        result.push(this.buffer[i]);
        i = (i + 1) % this.capacity;
      } while (i !== end);
    } else {
      // If not full, just go from head to tail
      while (i !== end) {
        result.push(this.buffer[i]);
        i = (i + 1) % this.capacity;
      }
    }

    return result;
  }

  /**
   * Gets the current number of items in the buffer.
   */
  size(): number {
    if (this.isFull) {
      return this.capacity;
    }
    if (this.tail >= this.head) {
      return this.tail - this.head;
    }
    return this.capacity - this.head + this.tail;
  }
}
