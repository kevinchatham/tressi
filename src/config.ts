import { promises as fs } from 'fs';
import path from 'path';
import { request } from 'undici';
import { z } from 'zod';

/**
 * Zod schema for a single request configuration.
 */
const RequestConfigSchema = z.object({
  /** The URL to send the request to. */
  url: z.string().url(),
  /** The request payload. Can be a JSON object or an array. */
  payload: z
    .record(z.string(), z.unknown())
    .or(z.array(z.unknown()))
    .optional(),
  /** The HTTP method to use for the request. Defaults to GET. */
  method: z
    .preprocess(
      (val) => (typeof val === 'string' ? val.toUpperCase() : val),
      z.enum(['GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'HEAD', 'OPTIONS']),
    )
    .default('GET'),
  /** Headers to be sent with this specific request. Merged with global headers. */
  headers: z.record(z.string(), z.string()).optional(),
});

/**
 * Zod schema for the main Tressi configuration.
 */
export const TressiConfigSchema = z.object({
  /** A URL to the JSON schema for this configuration file. */
  $schema: z.string().optional(),
  /** Global headers to be sent with every request. */
  headers: z.record(z.string(), z.string()).optional(),
  /** An array of request configurations. */
  requests: z.array(RequestConfigSchema),
});

/**
 * Type representing the Tressi configuration.
 */
export type TressiConfig = z.infer<typeof TressiConfigSchema>;

/**
 * Type representing a single request configuration.
 */
export type RequestConfig = z.infer<typeof RequestConfigSchema>;

/**
 * Loads and validates a Tressi configuration from a file, URL, or direct object.
 * @param configInput The path to a local config file, a URL to a remote config, or a config object.
 * @returns A promise that resolves to the validated Tressi configuration.
 */
export async function loadConfig(
  configInput: string | TressiConfig,
): Promise<TressiConfig> {
  if (typeof configInput === 'object') {
    return TressiConfigSchema.parse(configInput);
  }

  let rawContent: unknown;
  if (configInput.startsWith('http://') || configInput.startsWith('https://')) {
    const { statusCode, body } = await request(configInput);
    if (statusCode >= 400) {
      throw new Error(`Remote config fetch failed: ${statusCode}`);
    }
    rawContent = await body.json();
  } else {
    const absolutePath = path.resolve(configInput);
    const fileContent = await fs.readFile(absolutePath, 'utf-8');
    rawContent = JSON.parse(fileContent);
  }
  return TressiConfigSchema.parse(rawContent);
}
