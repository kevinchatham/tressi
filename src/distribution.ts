import chalk from 'chalk';

/**
 * Calculates the latency distribution for a given set of latencies.
 * @param latencies An array of latency values.
 * @param bucketCount The number of buckets to create.
 * @returns An array of bucket objects with range, count, and percentage values.
 */

export function getLatencyDistribution(
  latencies: number[],
  bucketCount = 6,
): {
  range: string;
  count: string;
  percent: string;
  cumulative: string;
}[] {
  if (latencies.length === 0) {
    return [];
  }

  const totalCount = latencies.length;
  const minLatency = Math.min(...latencies);
  const maxLatency = Math.max(...latencies);
  const range = maxLatency - minLatency;
  const bucketSize = Math.ceil(range / bucketCount) || 1;

  const buckets = Array.from({ length: bucketCount }, (_, i) => {
    const lowerBound = Math.floor(minLatency + i * bucketSize);
    const upperBound = Math.floor(minLatency + (i + 1) * bucketSize - 1);
    return {
      range: `${lowerBound}-${upperBound}`,
      count: 0,
      lowerBound,
      upperBound,
    };
  });

  // Make the last bucket catch all remaining values
  buckets[buckets.length - 1].upperBound = Infinity;
  buckets[buckets.length - 1].range =
    `${buckets[buckets.length - 1].lowerBound}+`;

  for (const latency of latencies) {
    const targetBucket = buckets.find(
      (b) => latency >= b.lowerBound && latency <= b.upperBound,
    );
    if (targetBucket) {
      targetBucket.count++;
    }
  }

  let cumulativeCount = 0;

  return buckets.map((bucket) => {
    const percentOfTotal =
      totalCount > 0 ? (bucket.count / totalCount) * 100 : 0;
    cumulativeCount += bucket.count;
    const cumulativePercent =
      totalCount > 0 ? (cumulativeCount / totalCount) * 100 : 0;

    return {
      range: bucket.range,
      count: bucket.count.toString(),
      percent: `${percentOfTotal.toFixed(1)}%`,
      cumulative: `${cumulativePercent.toFixed(1)}%`,
    };
  });
}
export function getStatusCodeDistribution(
  statusCodeMap: Record<number, number>,
): { code: string; count: string; percent: string; chart: string }[] {
  const totalCount = Object.values(statusCodeMap).reduce(
    (sum, count) => sum + count,
    0,
  );

  if (totalCount === 0) {
    return [];
  }

  const maxCount = Math.max(...Object.values(statusCodeMap));

  return Object.entries(statusCodeMap)
    .sort((a, b) => Number(a[0]) - Number(b[0]))
    .map(([code, count]) => {
      const percent = (count / totalCount) * 100;
      const chartBarCount =
        maxCount > 0 ? Math.round((count / maxCount) * 15) : 0;

      const numericCode = Number(code);
      let chartColor = chalk.green;
      if (numericCode >= 500) {
        chartColor = chalk.red;
      } else if (numericCode >= 400) {
        chartColor = chalk.red;
      } else if (numericCode >= 300) {
        chartColor = chalk.yellow;
      }

      const chart = chartColor('â–ˆ'.repeat(chartBarCount));

      return {
        code,
        count: count.toString(),
        percent: `${percent.toFixed(1)}%`,
        chart,
      };
    });
}
