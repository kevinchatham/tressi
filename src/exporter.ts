import chalk from 'chalk';
import { writeFile } from 'fs/promises';
import ora from 'ora';

import { RequestResult } from './stats';

export async function exportToCsv(path: string, results: RequestResult[]): Promise<void> {
  const writeSpinner = ora(`Writing CSV to ${path}`).start();
  try {
    const headers = ['url', 'status', 'latencyMs', 'success', 'error'];
    const rows = results.map((r) =>
      [
        `"${r.url}"`,
        r.status,
        r.latencyMs.toFixed(2),
        r.success,
        `"${r.error || ''}"`,
      ].join(',')
    );
    const csv = [headers.join(','), ...rows].join('\n');
    await writeFile(path, csv, 'utf-8');
    writeSpinner.succeed(`CSV saved to ${path}`);
  } catch (err) {
    writeSpinner.fail(chalk.red(`Failed to save CSV: ${(err as Error).message}`));
  }
} 