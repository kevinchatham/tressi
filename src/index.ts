import chalk from 'chalk';
import Table from 'cli-table3';
import { promises as fs } from 'fs';
import ora from 'ora';
import path from 'path';
import { z } from 'zod';
import { getLatencyDistribution } from './distribution';

import { loadConfig, RequestConfig, TressiConfig } from './config';
import { exportDataFiles } from './exporter';
import { Runner } from './runner';
import { average, RequestResult } from './stats';
import {
  generateMarkdownReport,
  generateSummary,
  TestSummary,
} from './summarizer';
import { TUI } from './ui';

export { RequestConfig, TestSummary, TressiConfig };

/**
 * Defines the options for a Tressi load test run.
 */
export interface RunOptions {
  /** The configuration for the test. Can be a path to a file, a URL, or a configuration object. */
  config: string | TressiConfig;
  /** The number of concurrent workers to use. Defaults to 10. For autoscale, this is the max workers. */
  workers?: number;
  /** The total duration of the test in seconds. Defaults to 10. */
  durationSec?: number;
  /** The time in seconds to ramp up to the target RPS. Defaults to 0. */
  rampUpTimeSec?: number;
  /** The target requests per second. If not provided, the test will run at maximum possible speed. */
  rps?: number;
  /** Whether to enable autoscale mode. Defaults to false. --rps is required for this. */
  autoscale?: boolean;
  /** The base path for the exported report. If not provided, no report will be generated. */
  exportPath?: string | boolean;
  /** Whether to use the terminal UI. Defaults to true. */
  useUI?: boolean;
  /** Suppress all console output. Defaults to false. */
  silent?: boolean;
}

/**
 * Prints a detailed summary of the load test results to the console.
 * @param results An array of `RequestResult` objects from the test run.
 * @param options The original `RunOptions` used for the test.
 */
function printSummary(
  results: RequestResult[],
  options: RunOptions,
  summary: TestSummary,
): void {
  const { global: globalSummary, endpoints: endpointSummaries } = summary;
  const {
    workers = 10,
    durationSec = 10,
    rps,
    rampUpTimeSec,
    autoscale,
  } = options;

  const configTable = new Table({
    head: ['Option', 'Setting', 'Argument'],
    colWidths: [20, 15, 20],
  });
  configTable.push([
    'Workers',
    autoscale ? `Up to ${workers}` : workers,
    '--workers',
  ]);
  configTable.push(['Duration', `${durationSec}s`, '--duration']);

  if (autoscale) {
    configTable.push(['Autoscale', 'Enabled', '--autoscale']);
  }

  if (rps) {
    configTable.push(['Target Req/s', rps, '--rps']);
  }

  if (rampUpTimeSec) {
    configTable.push(['Ramp-up Time', `${rampUpTimeSec}s`, '--ramp-up-time']);
  }

  // eslint-disable-next-line no-console
  console.log('\n' + chalk.bold('Run Configuration'));
  // eslint-disable-next-line no-console
  console.log(configTable.toString());

  const latencies = results.map((r) => r.latencyMs); // Keep for warning message logic

  if (rps) {
    const avgLatencyMs = average(latencies);
    if (avgLatencyMs > 0) {
      const maxRpsPerWorker = 1000 / avgLatencyMs;
      const maxPossibleRps = maxRpsPerWorker * workers;

      if (rps > maxPossibleRps) {
        const suggestedWorkers = Math.ceil(rps / maxRpsPerWorker);
        let warningMessage: string;

        if (autoscale) {
          warningMessage =
            `\n⚠️  Warning: Target of ${rps} Req/s may be unreachable.` +
            `\n   The autoscaler hit the maximum of ${workers} workers.` +
            `\n   With an average latency of ~${Math.ceil(
              avgLatencyMs,
            )}ms, the theoretical max is only ~${Math.floor(
              maxPossibleRps,
            )} Req/s.` +
            `\n   To meet the target, try increasing the --workers limit to at least ${suggestedWorkers}.`;
        } else {
          warningMessage =
            `\n⚠️  Warning: Target of ${rps} Req/s may be unreachable with ${workers} workers.` +
            `\n   With an average latency of ~${Math.ceil(
              avgLatencyMs,
            )}ms, the theoretical max is only ~${Math.floor(
              maxPossibleRps,
            )} Req/s.` +
            `\n   To meet the target, try increasing workers to at least ${suggestedWorkers}.`;
        }
        // eslint-disable-next-line no-console
        console.log(chalk.yellow(warningMessage));
      }
    }
  }

  const summaryTable = new Table({
    head: ['Stat', 'Value'],
    colWidths: [30, 20],
  });

  summaryTable.push(
    ['Duration', `${Math.ceil(globalSummary.duration)}s`],
    ['Total Requests', globalSummary.totalRequests],
    [chalk.green('Successful'), globalSummary.successfulRequests],
    [chalk.red('Failed'), globalSummary.failedRequests],
  );

  if (rps && globalSummary.theoreticalMaxRps) {
    summaryTable.push(
      [
        'Req/s (Actual/Target)',
        `${Math.ceil(globalSummary.actualRps)} / ${rps}`,
      ],
      [
        'Req/m (Actual/Target)',
        `${Math.ceil(globalSummary.actualRps * 60)} / ${rps * 60}`,
      ],
      ['Theoretical Max Req/s', globalSummary.theoreticalMaxRps.toFixed(0)],
      ['Achieved %', `${globalSummary.achievedPercentage.toFixed(0)}%`],
    );
  } else {
    summaryTable.push(
      ['Req/s', Math.ceil(globalSummary.actualRps)],
      ['Req/m', Math.ceil(globalSummary.actualRps * 60)],
    );
  }

  summaryTable.push(
    ['Avg Latency', `${Math.ceil(globalSummary.avgLatencyMs)}ms`],
    ['p95 Latency', `${Math.ceil(globalSummary.p95LatencyMs)}ms`],
    ['p99 Latency', `${Math.ceil(globalSummary.p99LatencyMs)}ms`],
  );

  // eslint-disable-next-line no-console
  console.log('\n' + chalk.bold('Global Test Summary'));
  // eslint-disable-next-line no-console
  console.log(summaryTable.toString());

  const latencyDistribution = getLatencyDistribution(
    results.map((r) => r.latencyMs),
  );

  if (latencyDistribution.length > 0) {
    const distributionTable = new Table({
      head: ['Range (ms)', 'Count', '% Total', 'Cumulative', 'Chart'],
      colWidths: [15, 10, 10, 12, 20],
    });

    latencyDistribution.forEach((bucket) => {
      distributionTable.push([
        bucket.range,
        bucket.count,
        bucket.percent,
        bucket.cumulative,
        bucket.chart,
      ]);
    });

    // eslint-disable-next-line no-console
    console.log(chalk.bold('\nLatency Distribution'));
    // eslint-disable-next-line no-console
    console.log(distributionTable.toString());
  }

  const statusCodeMap: Record<number, number> = results.reduce(
    (acc, r) => {
      acc[r.status] = (acc[r.status] || 0) + 1;
      return acc;
    },
    {} as Record<number, number>,
  );

  const statusTable = new Table({
    head: ['Status Code', 'Count'],
    colWidths: [20, 20],
  });

  Object.entries(statusCodeMap)
    .sort((a, b) => parseInt(a[0]) - parseInt(b[0]))
    .forEach(([code, count]) => {
      const color = code.startsWith('2')
        ? chalk.green
        : code.startsWith('5')
          ? chalk.red
          : chalk.yellow;
      statusTable.push([color(code), count]);
    });

  if (statusTable.length > 0) {
    // eslint-disable-next-line no-console
    console.log(chalk.bold('\nResponses by Status Code:'));
    // eslint-disable-next-line no-console
    console.log(statusTable.toString());
  }

  if (endpointSummaries.length > 1) {
    // eslint-disable-next-line no-console
    console.log('\n' + chalk.bold('Summary by Endpoint'));
    endpointSummaries.forEach((endpointSummary) => {
      const {
        url,
        totalRequests,
        successfulRequests,
        failedRequests,
        avgLatencyMs,
        p95LatencyMs,
        p99LatencyMs,
      } = endpointSummary;
      const endpointTable = new Table({ colWidths: [30, 60] });
      endpointTable.push({ Endpoint: chalk.cyan(url) });
      endpointTable.push({ 'Total Requests': totalRequests });
      endpointTable.push({ Successful: chalk.green(successfulRequests) });
      endpointTable.push({ Failed: chalk.red(failedRequests) });
      endpointTable.push({
        'Avg Latency': `${Math.ceil(avgLatencyMs)}ms`,
      });
      endpointTable.push({
        'p95 Latency': `${Math.ceil(p95LatencyMs)}ms`,
      });
      endpointTable.push({
        'p99 Latency': `${Math.ceil(p99LatencyMs)}ms`,
      });

      // eslint-disable-next-line no-console
      console.log(endpointTable.toString());
    });
  }

  const errors = results.filter((r) => r.error).slice(0, 5);
  if (errors.length > 0) {
    // eslint-disable-next-line no-console
    console.log(chalk.bold.red('\nSampled Errors:'));
    errors.forEach((e) => {
      // eslint-disable-next-line no-console
      console.log(`- ${e.url}: ${e.error}`);
    });
  }
}

/**
 * The main function to execute a Tressi load test.
 * It loads the configuration, initializes the runner, starts the UI,
 * and prints a summary upon completion.
 * @param options The `RunOptions` for the test.
 * @returns A `Promise` that resolves with the `TestSummary` object.
 */
export async function runLoadTest(options: RunOptions): Promise<TestSummary> {
  const { silent = false, useUI = true } = options;
  const spinner = ora({
    text: 'Loading config...',
    isEnabled: !silent,
  }).start();
  let loadedConfig: TressiConfig;
  try {
    loadedConfig = await loadConfig(options.config);
    spinner.succeed(`Loaded ${loadedConfig.requests.length} request targets`);
  } catch (err) {
    if (err instanceof z.ZodError) {
      spinner.fail('Config validation failed:');
      if (!silent) {
        // eslint-disable-next-line no-console
        console.error(JSON.stringify(err.errors, null, 2));
      }
    } else {
      spinner.fail(`Failed to load config: ${(err as Error).message}`);
    }
    throw err;
  }

  const runner = new Runner(
    options,
    loadedConfig.requests,
    loadedConfig.headers || {},
  );

  // If we have a TUI, we need to handle its destruction and polling
  if (useUI && !silent) {
    const tui = new TUI(() => runner.stop());
    const tuiInterval = setInterval(() => {
      const latencies = runner.getLatencies();
      const statusCodes = runner.getStatusCodeMap();
      const startTime = runner.getStartTime();
      const elapsedSec = startTime > 0 ? (Date.now() - startTime) / 1000 : 0;
      const totalSec = options.durationSec || 10;

      tui.update(
        latencies,
        statusCodes,
        runner.getCurrentRps(),
        elapsedSec,
        totalSec,
        options.rps,
        runner.getSuccessfulRequestsCount(),
        runner.getFailedRequestsCount(),
        runner.getAverageLatency(),
        runner.getWorkerCount(),
      );
    }, 500);

    runner.on('stop', () => {
      clearInterval(tuiInterval);
      tui?.destroy();
    });
  } else {
    // If we're not using the TUI, we should still provide some basic feedback
    const noUiSpinner = ora({
      text: 'Test starting...',
      isEnabled: !silent,
    }).start();
    const noUiInterval = setInterval(() => {
      const startTime = runner.getStartTime();
      const elapsedSec = startTime > 0 ? (Date.now() - startTime) / 1000 : 0;
      const totalSec = options.durationSec || 10;

      const rps = runner.getCurrentRps();
      const successful = runner.getSuccessfulRequestsCount();
      const failed = runner.getFailedRequestsCount();
      const workers = runner.getWorkerCount();
      const avgLatency = runner.getAverageLatency();

      const rpsDisplay = options.rps ? `${rps}/${options.rps}` : `${rps}`;

      noUiSpinner.text = `[${elapsedSec.toFixed(0)}s/${totalSec}s] Req/s: ${
        rpsDisplay
      } | Workers: ${workers} | Success: ${successful} | Fail: ${failed} | Avg Latency: ${avgLatency.toFixed(
        0,
      )}ms`;
    }, 1000);

    runner.on('stop', () => {
      clearInterval(noUiInterval);
      noUiSpinner.succeed('Test finished. Generating summary...');
    });
  }

  const results = await runner.run();
  const startTime = runner.getStartTime();
  const actualDurationSec = startTime > 0 ? (Date.now() - startTime) / 1000 : 0;
  const summary = generateSummary(results, options, actualDurationSec);

  if (options.exportPath) {
    const exportSpinner = ora({
      text: 'Exporting results...',
      isEnabled: !silent,
    }).start();
    try {
      const baseExportName =
        typeof options.exportPath === 'string'
          ? options.exportPath
          : 'tressi-report';
      const runDate = new Date();
      const reportDir = path.resolve(
        process.cwd(),
        `${baseExportName}-${runDate.toISOString()}`,
      );
      await fs.mkdir(reportDir, { recursive: true });

      const markdownReport = generateMarkdownReport(
        summary,
        options,
        results,
        loadedConfig,
        {
          exportName: baseExportName,
          runDate,
        },
      );
      await fs.writeFile(path.join(reportDir, 'report.md'), markdownReport);

      await exportDataFiles(summary, results, reportDir);

      exportSpinner.succeed(`Successfully exported results to ${reportDir}`);
    } catch (err) {
      exportSpinner.fail(
        chalk.red(`Failed to export results: ${(err as Error).message}`),
      );
    }
  }

  // Final summary to console
  if (!silent) {
    printSummary(results, options, summary);
  }

  return summary;
}
