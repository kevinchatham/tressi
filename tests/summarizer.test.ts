import { build, Histogram } from 'hdr-histogram-js';
import { afterAll, beforeAll, describe, expect, it, vi } from 'vitest';

import { RunOptions } from '../src';
import { TressiConfig } from '../src/config';
import { Distribution } from '../src/distribution';
import { Runner } from '../src/runner';
import { generateMarkdownReport, generateSummary } from '../src/summarizer';

const createHistogram = (latencies: number[]): Histogram => {
  const histogram = build();
  for (const l of latencies) {
    histogram.recordValue(l);
  }
  return histogram;
};

const mockRunner = {
  getSampledResults: () => [
    {
      method: 'GET',
      url: 'http://a.com',
      latencyMs: 100,
      status: 200,
      success: true,
      timestamp: 1,
    },
    {
      method: 'GET',
      url: 'http://a.com',
      latencyMs: 150,
      status: 200,
      success: true,
      timestamp: 2,
    },
    {
      method: 'GET',
      url: 'http://b.com',
      latencyMs: 200,
      status: 200,
      success: true,
      timestamp: 3,
    },
    {
      method: 'GET',
      url: 'http://b.com',
      latencyMs: 500,
      status: 500,
      success: false,
      error: 'Error',
      timestamp: 4,
    },
  ],
  getHistogram: () => createHistogram([100, 150, 200, 500]),
  getEndpointHistograms: () => {
    const map = new Map<string, Histogram>();
    map.set('GET http://a.com', createHistogram([100, 150]));
    map.set('GET http://b.com', createHistogram([200, 500]));
    return map;
  },
  getStatusCodeMap: () => ({ 200: 3, 500: 1 }),
  getSuccessfulRequestsCount: () => 3,
  getFailedRequestsCount: () => 1,
  getAverageLatency: () => 237.5,
  getDistribution: () => {
    const distribution = new Distribution();
    [100, 150, 200, 500].forEach((l) => distribution.add(l));
    return distribution;
  },
  getSuccessfulRequestsByEndpoint: () => {
    const map = new Map<string, number>();
    map.set('GET http://a.com', 2);
    map.set('GET http://b.com', 1);
    return map;
  },
  getFailedRequestsByEndpoint: () => {
    const map = new Map<string, number>();
    map.set('GET http://a.com', 0);
    map.set('GET http://b.com', 1);
    return map;
  },
} as unknown as Runner;

const mockOptions: RunOptions = {
  config: { requests: [] },
  durationSec: 10,
};

const mockConfig: TressiConfig = {
  requests: [{ url: 'http://a.com', method: 'GET' }],
};

/**
 * Test suite for the summary generation logic.
 */
describe('summarizer', () => {
  beforeAll(() => {
    vi.useFakeTimers();
    vi.setSystemTime(new Date('2023-01-01T03:00:00.000Z'));
  });

  afterAll(() => {
    vi.useRealTimers();
  });

  /**
   * It should correctly calculate all global statistics for a given set of results.
   */
  it('should generate an accurate global summary', () => {
    const summary = generateSummary(mockRunner, mockOptions, 10);
    const { global: g } = summary;

    expect(g.totalRequests).toBe(4);
    expect(g.successfulRequests).toBe(3);
    expect(g.failedRequests).toBe(1);
    expect(g.avgLatencyMs).toBeCloseTo(237.5);
    expect(g.minLatencyMs).toBe(100);
    expect(g.maxLatencyMs).toBe(500);
    expect(g.p95LatencyMs).toBe(500);
    expect(g.actualRps).toBe(0.4);
    expect(g.theoreticalMaxRps).toBe(0);
    expect(g.achievedPercentage).toBe(0);
  });

  /**
   * It should correctly aggregate results by URL and calculate statistics for each endpoint.
   */
  it('should generate an accurate summary for each endpoint', () => {
    const summary = generateSummary(mockRunner, mockOptions);
    const { endpoints: e } = summary;

    expect(e).toHaveLength(2);

    const summaryA = e.find((s) => s.url === 'http://a.com');
    expect(summaryA).toBeDefined();
    expect(summaryA?.method).toBe('GET');
    expect(summaryA?.totalRequests).toBe(2);
    expect(summaryA?.successfulRequests).toBe(2);
    expect(summaryA?.failedRequests).toBe(0);
    expect(summaryA?.avgLatencyMs).toBe(125);
    expect(summaryA?.p95LatencyMs).toBe(150);

    const summaryB = e.find((s) => s.url === 'http://b.com');
    expect(summaryB).toBeDefined();
    expect(summaryB?.method).toBe('GET');
    expect(summaryB?.totalRequests).toBe(2);
    expect(summaryB?.successfulRequests).toBe(1);
    expect(summaryB?.failedRequests).toBe(1);
    expect(summaryB?.avgLatencyMs).toBe(350);
    expect(summaryB?.p95LatencyMs).toBe(500);
  });

  /**
   * It should generate a Markdown report containing all summary sections.
   */
  it('should generate a comprehensive Markdown report', () => {
    const summary = generateSummary(mockRunner, mockOptions, 10);
    const metadata = {
      exportName: 'my-test-report',
      runDate: new Date('2022-12-31T21:00:00.000Z'),
    };
    const markdown = generateMarkdownReport(
      summary,
      mockOptions,
      mockRunner,
      mockConfig,
      metadata,
    );

    // Check that the report contains expected sections and data
    expect(markdown).toContain('# Tressi Load Test Report');
    expect(markdown).toContain('| Metric | Value |');
    expect(markdown).toContain('Version |');
    expect(markdown).toContain('Export Name | my-test-report');
    expect(markdown).toContain('Test Time |');
    expect(markdown).toContain('## Analysis & Warnings');
    expect(markdown).toContain('## Global Summary');
    expect(markdown).toContain('## Endpoint Summary');
    expect(markdown).toContain('GET http://a.com');
    expect(markdown).toContain('GET http://b.com');
  });
});
