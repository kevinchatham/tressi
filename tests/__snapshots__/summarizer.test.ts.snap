// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`summarizer > should generate a comprehensive Markdown report 1`] = `
"# Tressi Load Test Report

| Metric | Value |
|---|---|
| Version | 0.0.8 |
| Export Name | my-test-report |
| Test Time | 12/31/2022, 9:00:00 PM |

## Analysis & Warnings ⚠️

> *This section highlights potential performance issues or configuration problems detected during the test.*

* **Target RPS Unreachable**: The target of 1 RPS was not met. The test achieved ~0 RPS (40% of the target). Based on the average latency of 238ms, you would need at least **2** workers to meet the target.
* **High Failure Rate**: The endpoint \`http://b.com\` had a failure rate of 50.0%. This may indicate a problem under load.

<details>
<summary>View Full Test Configuration</summary>

\`\`\`json
{
  "requests": [
    {
      "url": "http://a.com",
      "method": "GET"
    }
  ]
}
\`\`\`

</details>

## Run Configuration

> *This table shows the main parameters used for the load test run.*

| Option | Setting | Argument |
|---|---|---|
| Workers | 10 | \`--workers\` |
| Duration | 10s | \`--duration\` |
| Target Req/s | 1 | \`--rps\` |
## Global Summary

> *A high-level overview of the entire test performance across all endpoints.*

| Stat | Value |
| --- | --- |
| Duration | 10s |
| Total Requests | 4 |
| Successful | 3 |
| Failed | 1 |
| Req/s (Actual/Target) | 0 / 1 |
| Req/m (Actual/Target) | 24 / 60 |
| Theoretical Max Req/s | 1 |
| Achieved % | 40% |
| Avg Latency | 238ms |
| Min Latency | 100ms |
| Max Latency | 500ms |
| p95 Latency | 500ms |
| p99 Latency | 500ms |

## Latency Distribution

> *This table shows how request latencies were distributed. **% of Total** is the percentage of requests that fell into that specific time range. **Cumulative %** is the running total, showing the percentage of requests at or below that latency.*

| Range (ms) | Count | % of Total | Cumulative % | Chart |
|---|---|---|---|---|
| 100-149ms | 1 | 25% | 25% | ████████████████████ |
| 150-199ms | 1 | 25% | 50% | ████████████████████ |
| 200-249ms | 1 | 25% | 75% | ████████████████████ |
| 450+ms | 1 | 25% | 100% | ████████████████████ |

## Error Summary

> *This section summarizes any errors that occurred during the test, grouped by message.*

| Count | Error Message |
|---|---|
| 1 | Error |

## Responses by Status Code

> *A breakdown of all responses by their HTTP status code.*

| Status Code | Count |
|---|---|
| 200 | 3 |
| 500 | 1 |

## Endpoint Summary

> *A summary of request outcomes for each endpoint.*

| Endpoint | Success | Failed |
|---|---|---|
| GET http://a.com | 2 | 0 |
| GET http://b.com | 1 | 1 |

## Endpoint Latency

> *A detailed latency breakdown for each individual API endpoint.*

| Endpoint | Avg | Min | Max | P95 | P99 |
|---|---|---|---|---|---|
| GET http://a.com | 125ms | 100ms | 150ms | 150ms | 150ms |
| GET http://b.com | 350ms | 200ms | 500ms | 500ms | 500ms |
"
`;
